scVI

sjfleming opened the pull request:
Implement scVI in Cellarium ML.

sjfleming committed:
Yang's foundation for scVI implementation from scvi-tools (#130)

Initial scVI code port

sjfleming committed:
Hackathon progress (#133)

* partially there

* model that will run from command line

sjfleming committed:
scVI example config file (#134)

* config file example

* remove obsolete gitignore

sjfleming committed:
remove obsolete gitignore

mbabadi commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,521 @@
+        use_observed_lib_size:
+            If ``True``, use the observed library size for RNA as the scaling factor in the mean of the
+            conditional distribution.
+        library_log_means:
```
These can be computed via a separate one-pass model, `OnePassLibrarySizeStatistics`

mbabadi commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,521 @@
+            set to ``"embedding"``.
+    """
+
+    def __init__(
```
Additional flexibility that is needed:

- Which layer of the encoder and/or decoder to include batch variables; "-1" would mean nowhere!
- The option to either learn batch variables, to use "batch vectors" that are optionally provided to the initializer

mbabadi commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,521 @@
+            **_extra_decoder_kwargs,
+        )
+
+    def reset_parameters(self) -> None:
```
If reset_parameters is over-ridden with `pass`, why does it work?

mbabadi commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,521 @@
+
+        return {"loss": loss}
+    
+    def predict(
```
Implement

mbabadi commented on cellarium/ml/models/common/decoder.py:
```diff
@@ -0,0 +1,179 @@
+from cellarium.ml.models.common.fclayer import FCLayers
+
+
+class DecoderSCVI(nn.Module):
```
As a matter of principle, things that belong to `common` must be common :D `DecoderSCVI` is not common, because it is only used by SCVI. Either refactor it to something more general-purpose, or refactor out to `scvi.py`.

mbabadi commented on cellarium/ml/models/common/distributions.py:
```diff
@@ -0,0 +1,189 @@
+    log = log_fn
+    lgamma = lgamma_fn
+    log_theta_mu_eps = log(theta + mu + eps)
+    res = (
```
This is not robust when theta is large, see cellarium-gpt branch.

mbabadi commented on cellarium/ml/models/common/encoder.py:
```diff
@@ -0,0 +1,113 @@
+    return x
+
+
+class Encoder(nn.Module):
```
Same comment as earlier about `decoder.py` not being general enough to be under `common`.

mbabadi reviewed (commented):
General comment: let's also make a `test_scvi.py` and include an scvi test under `test_cli.py`.

sjfleming committed:
Fix a major typo in loss computation

sjfleming committed:
black and ruff code formatting

ImXman reviewed (commented):
reviewed

ImXman committed:
Implement reset_parameters() for scvi (#151)

* reset_params

Added reset_parameters.
1) reset_parameters specifies how to intialize model.
2) An updated config yaml is added to exmple_configs/

* Update scvi.py

* Update scvi.py

* Update scvi.py

corrected indentation

* Update scvi.py

added arg doc for predict

---------

Co-authored-by: Yang Xu <yxu@broadinstitute.org>

sjfleming committed:
Refactor the negative binomial code (#150)

* Yerdos NB implementation

* Create a mixin to log learning rate

* scVI demo notebook

sjfleming committed:
Refactor common model code (#155)

* Refactor

sjfleming committed:
scvi: make encoder/decoder more flexible (#156)

* Some helper classes

* Revert the LogLearningRateMixin - use callback

* Work on encoder

* Encoder potentially seems to work

* Decoder seems to be working

* Change common filename and cleanup

* Bug where VAE was not using norm and dropout args

* Remove defunct inputs, speed up using 4 workers

* Change back to an scvi-tools computation of batch biases

* Switch model to use batch indices rather than bias Parameters

* New example config for new setup

* Clean up docstring

* No dropout in the decoder

* Modify the reset_parameters() function

ImXman committed:
reset_parameters update (#158)

* reset_parameters update

All parameters should be initialized the same time when calling reset_parameters()

* Update scvi.py

corrected a typo

---------

Co-authored-by: Yang Xu <yxu@broadinstitute.org>

sjfleming committed:
Revert "reset_parameters update (#158)" (#159)

This reverts commit 196820cab1f504dbd3dbe9324aef5b1bcba6d4fb.

sjfleming committed:
Add scvi test to test_cli.py (#162)

sjfleming committed:
Add second test data shard to scvi cli test (after #161)

sjfleming committed:
scvi - enable batch embeddings, sampling, and freezing of batch bias computations (#169)

* Refactor LinearWithBatch to take batch embeddings rather than indices

* Batch embeddings; can be random variables; can have KL

* Refactor VAE config; enable affine transformation in final decoder layer

sjfleming committed:
Yang's starting point plus small fixes (#176)

sjfleming committed:
Fix bugs found during testing (#177)

Fix saving and loading among other things

sjfleming committed:
Refactor negative binomial and add .reconstruct() method (#178)

sjfleming committed:
Add some hints to config

sjfleming committed:
Commit messy de-batched example notebook

sjfleming committed:
Fix NaN in scvi (#215)

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
Implement categorical covariates (#254)

sjfleming committed:
Remove Parameters when not used

sjfleming committed:
Attempt to fix test failure

sjfleming committed:
Another attempt to fix test failure

sjfleming committed:
Categorical covariates were missing from final_layer of encoder

ordabayevy commented on cellarium/ml/models/common/gather.py:
Remove this file?

sjfleming replied:
Oops, yes, thanks

ordabayevy commented on cellarium/ml/models/common/nn.py:
This can be moved to a new `cellarium.ml.layers` sub-module.

sjfleming replied:
Sounds good

ordabayevy commented on tests/test_cli.py:
```diff
@@ -1,9 +1,14 @@
 # Copyright Contributors to the Cellarium project.
 # SPDX-License-Identifier: BSD-3-Clause
 
+from __future__ import annotations
```
I have though about using `from __future__ import annotations` and adding `if TYPE_CHECKING` clause ([PR for it](https://github.com/cellarium-ai/cellarium-ml/pull/248)). However, it seems that the future of `from __future__ import annotations` is uncertain and it might get deprecated. What this import does is it stringifies annotations, however, there was a backlash against it and a new method was proposed that will defer annotations (PEP 649). This new method is expected to be released next year.

ordabayevy replied:
Yeah, I think it might be better to just wait for the new method and implement it once it is realeased.

sjfleming committed:
Remove obsolete file, move nn code to layers submodule

sjfleming committed:
Revert use of TYPE_CHECKING

sjfleming committed:
Code cleanup

sjfleming committed:
scvi predict; notebook demo

sjfleming committed:
Fix config typo

sjfleming committed:
Allow batch to be specified by multiple categorical columns (has drawbacks)

ordabayevy commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,981 @@
+            encoder_input_ng = torch.log1p(encoder_input_ng)
+
+        # if continuous_covariates_nc is not None and self.encode_covariates:
+        #     encoder_input_ng = torch.cat((encoder_input_ng, continuous_covariates_nc), dim=-1)
```
Remove comment lines

ordabayevy commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,981 @@
+                raise NotImplementedError
+                # px_r = linear(
+                #     torch.nn.functional.one_hot(y.squeeze().long(), self.n_labels).float(), self.px_r
+                # )  # px_r gets transposed - last dimension is nb genes
```
Remove comment lines

ordabayevy commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,981 @@
+        DataFrame with columns as batch covariates and an extra column "scvi_batch_code" with the
+        code used in the model.
+    """
+    print("WARNING: The batch_index_to_batch_label lookup for multiple batch_keys is still experimental.")
```
Use actual warning

ordabayevy reviewed (commented):
Have you considered using Pyro to write the model? I wonder whether it would simplify the code.

sjfleming committed:
Two config files for a big run of scvi

sjfleming committed:
gitignore data files

sjfleming committed:
Extra config files

sjfleming committed:
Small modifications

sjfleming committed:
Notebook for highly variable genes

sjfleming committed:
Merge branch 'modified_scvi' of https://github.com/cellarium-ai/cellarium-ml into modified_scvi

sjfleming committed:
PredictionWriter and config files

sjfleming committed:
Modifications for reconstruction

sjfleming committed:
typechecking

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
make format

sjfleming committed:
mypy

sjfleming committed:
make format

sjfleming committed:
scvi unit tests comparing to scvi-tools (#317)

* add unit test for scvi encoder

* make format

* mypy

* add decoder unit test

* full scvi-tools testing works

sjfleming committed:
scvi: implement KL annealing, tons of tests, and fix a bug (#328)

* KL weight annealing

* scvi-tools parity tests

* important fix: use latent mean do not sample

* add reconstruction capability with tests

sjfleming committed:
mypy fix

sjfleming committed:
update test file paths

sjfleming committed:
update test to read files from http

sjfleming committed:
scvi: add tests for categorical covariates and reconstructions (#338)

* added categorical covariates to test of gradients and training dynamics

* test that reconstructed counts match

* updates to reconstruction test

* fix error with categorical covariate handling

* fixed some reconstruction scaling issues, tests fail still

* simulated reconstruction test passes

* all tests passing

* format and typecheck

* turn KL max weight 10 test back on

sjfleming committed:
Update reconstruction example

sjfleming committed:
fix gene labeling error

sjfleming committed:
check requested reconstruction genes

sjfleming committed:
address test flakiness somewhat

sjfleming committed:
test reconstructions with transform_batch=None
Fixes #356

sjfleming committed:
scvi: allow injection of categorical covariates without batch (#345)

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
update arg name in scvi cli test

sjfleming committed:
address review comments

sjfleming committed:
resolving some issues with using supplied library size (not recommended)

sjfleming committed:
address test flakiness

sjfleming committed:
scvi: add optional normalizing flow prior and fix reconstruction bug (#358)

* normalizing flow prior with zuko

* demo notebooks

* important fixes to reconstruction, especially with categorical covariates

sjfleming committed:
fix logic to allow DDP to work with several devices

sjfleming committed:
remove obsolete assertion

sjfleming committed:
add series_to_str_list for use by FileLoader

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
remove redundant function

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
remove extraneous files

sjfleming committed:
remove extra notebooks

sjfleming committed:
scvi-tools seems to be missing a zarr dependency

fedorgrab commented on cellarium/ml/layers/vae.py:
```diff
@@ -0,0 +1,104 @@
+        in_features: int,
+        out_features: int,
+        n_hidden: list[int],
+        dressing_init_kwargs: dict[str, Any] = {},
```
Here's a default value set to mutable too 😅

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+        out_features: int,
+        label_to_bias_hidden_layers: list[int],
+        n_batch: int = 0,
+        categorical_covariate_dimensions: list[int] = [],
```
default mutable

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+        n_batch: int = 0,
+        categorical_covariate_dimensions: list[int] = [],
+        bias: bool = True,
+        label_to_bias_dressing_init_kwargs: dict[str, Any] = {},
```
default mutable

fedorgrab commented on cellarium/ml/layers/vae.py:
```diff
@@ -0,0 +1,104 @@
+    ):
+        assert not (use_batch_norm and use_layer_norm), "Cannot use both batch and layer normalization."
+        super().__init__()
+        out_features = layer.out_features
```
nit:
```python3
out_features = getattr(layer, "out_features", None)
if out_features is None:
    raise ValueError("Layer must define out_features")
```

sjfleming replied:
thanks that's good

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+        n_batch: int = 0,
+        n_latent: int = 10,
+        n_continuous_cov: int = 0,
+        n_cats_per_cov: list[int] = [],
```
default mutable

fedorgrab commented on pyproject.toml:
```diff
@@ -34,6 +34,7 @@ dependencies = [
   "scikit-misc==0.5.1",
   "torch>=2.2.0",
   "transformers",
+  "zuko",
```
nit: pin dependency

fedorgrab commented on pyproject.toml:
```diff
@@ -47,6 +48,8 @@ test = [
   "tenacity",
   "tensorboard",
   "torchvision",
+  "scvi-tools",
```
nit: pin dependency

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+            case "gene-label":
+                inverse_overdispersion = None
+                raise NotImplementedError
+        assert isinstance(inverse_overdispersion, torch.Tensor)
```
in case of "gene-cell" this will raise an error. Because decoder can take None for `inverse_overdispersion`, but it would have to have `inverse_overdispersion_encoder`, which will be None. If in Decoder init there will be non-Poisson gene-likelihood + `gene-cell` dispersion, it will create `inverse_overdispersion_encoder`. I think maybe this assert is redundant, because basically seems like `inverse_overdispersion` can be `None` (according to Decoder).

sjfleming replied:
I think you're right. I removed the assertion. I have never tested with anything but "gene" overdispersion to be honest, and I don't think it's covered by tests. We can leave this as a todo. I think most people run scVI with "gene" overdispersion, so it was never a priority, but the code was in scvi-tools so we wanted to include it... we just did a half job

fedorgrab commented on cellarium/ml/models/scvi.py:
I think it would be a good idea to restructure this file. I see it as a package `scvi`, instead of one file. It could look like:
```
models/scvi/
├── __init__.py              # Re-exports everything at current paths
├── model.py                 # SingleCellVariationalInference (~600 lines)
├── layers.py                # LinearWithStructuredBias, LinearWithBatch,
│                            #   LinearWithBatchAndCovariates, LinearWithCovariates,
│                            #   FullyConnectedWithBatchArchitecture (~150 lines)
├── encoder.py               # EncoderSCVI (~70 lines)
├── decoder.py               # DecoderSCVI (~100 lines)
└── utils.py                 # class_from_class_path, instantiate_from_class_path,
                             #   weights_init, compute_annealed_kl_weight,
                             #   batch_index_to_batch_label,
                             #   enumerate_observed_batch_covariate_combinations,
                             #   _enumerate_categorical_combinations,
                             #   _n_cats_per_column (~200 lines)
```

fedorgrab replied:
Maybe encoder and decoder could be one file.

fedorgrab replied:
Maybe they even could be part of scvi/layers.

sjfleming replied:
this is a good idea, but I will leave the refactor for a separate issue

sjfleming replied:
it's now #394

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+            in_features=self.fully_connected.out_features,
+            out_features=out_features,
+            bias=final_layer["init_args"].pop("bias", True),
+            **final_layer["init_args"],
```
I think this one is critical (unless me and copilot are missing some context)

sjfleming replied:
it was not technically a problem since I only ever used that final_layer dict here, but I've changed it to make the intent more explicit and I've adpoted an approach that does not mutate the dict

fedorgrab reviewed (changes requested):
I am still learning the codebase, so I didn't review it deeply, but it was a good dive in to scvi! I (w/ copilot) left couple of suggestions. In general, exciting stuff!

One more general suggestion, I would change all new asserts to a proper check and raise ValueError, or RuntimeError. I think some of the code in this branch is outdated, and some asserts are leftovers from earlier stages of the repository. At some point, we switched to using explicit error-raising validation checks in new code, so I assume these asserts are from older code.

fedorgrab commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+                * ``"decoder"``: use layer norm only in the decoder.
+                * ``"both"``: use layer norm in both encoder(s) and decoder.
+            Note: if ``use_batch_norm`` is also specified, both will be applied (first
+            :class:`~torch.nn.BatchNorm1d`, then :class:`~torch.nn.LayerNorm`).
```
This one seems to be a helpful comment

sjfleming replied:
fixed

sjfleming committed:
scvi: add validation method and metrics (#380)

* optionally compute and log validation metrics during training

sjfleming committed:
add val_reconstruction_loss (#381)

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming committed:
Merge branch 'main' into modified_scvi

sjfleming commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+                        importance_weight_means.append(importance_weight_n1)
+
+                    # run generative model and scale output
+                    scaled_counts_np = _run_generative_and_scale_output(
+                        z_nk=sampled_z_nk,
+                        local_batch_nb=local_batch_nb,
+                        local_categorical_covariate_np=local_categorical_covariate_np,
+                    )
+                    output_counts_sum_np += scaled_counts_np * importance_weight_n1
+
+                x_tilde_np = output_counts_sum_np / (len(transformed_batch_index_n_list) * sum(importance_weight_means))
+
```
a valid comment, identifying a real problem with averaging over multiple transformed batches (maybe this is why it never worked well...). addressed it.

sjfleming committed:
ignore claude setup

sjfleming committed:
fix mean multi-batch reconstruction code path

sjfleming commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+        # construct the count distribution
+        dist: Distribution
+        match self.gene_likelihood:
+            case "nb":
```
valid comment, addressed

sjfleming commented on cellarium/ml/models/scvi.py:
```diff
@@ -0,0 +1,1469 @@
+                        typically referred to as ``z_nk``.
+                - If :attr:`self.reconstruct_counts_on_predict` is False:
+                    - (the g in x_ng is a notational misnomer) Reconstruction of the input data: ``x_ng'``, which
+                        will have a different number of genes depending on :attr:`self.reconstruction_var_names_g`.
```
glad somebody is reading these docstrings

sjfleming commented on tests/test_scvi.py:
```diff
@@ -0,0 +1,2559 @@
+            "init_args": {"out_features": 32, "label_to_bias_hidden_layers": []},
+            "final_additive_bias": True,
+        },
+    ]
```
this is true

sjfleming committed:
responses to review

sjfleming committed:
added optional dropout on counts input to encoder at train time

sjfleming committed:
attempt to speed up cli tests and reduce test load

sjfleming committed:
address scanpy dataset loading flakiness in test

sjfleming committed:
parallelize CI tests for speed
